sklearn/cluster/hierarchical.py

```diff
@@ -713,7 +713,7 @@ def fit(self, X, y=None):
         -------
         self
         """
-        X = check_array(X)
+        X = check_array(X, ensure_min_samples=2)
```

Comment:
I checked clustering with `n_clusters > n_samples` and we already get a good error message from line 586 (in master).

On master also, clustering 1 sample with 1 cluster crashes with:

```
>>> AgglomerativeClustering(n_clusters=1).fit(X[0:1])
Traceback (most recent call last):
  File "<ipython-input-8-852a66749455>", line 1, in <module>
    AgglomerativeClustering(n_clusters=1).fit(X[0:1])
  File "/volatile/ogrisel/code/scikit-learn/sklearn/cluster/hierarchical.py", line 769, in fit
    self.n_leaves_)
  File "/volatile/ogrisel/code/scikit-learn/sklearn/cluster/hierarchical.py", line 592, in _hc_cut
    nodes = [-(max(children[-1]) + 1)]
IndexError: index -1 is out of bounds for axis 0 with size 0
```

I am fine with setting `ensure_min_samples=2` as a stopgap in this PR. At least it's better than the state of master.